perl.c

```diff
@@ -4227,6 +4227,19 @@ S_open_script(pTHX_ const char *scriptname, bool dosearch, bool *suidscript)
                 Safefree(PL_origfilename);
                 PL_origfilename = (char *)scriptname;
             }
+            else {
+                char proc_fd_path[64];
+                snprintf(proc_fd_path, sizeof(proc_fd_path), "/proc/self/fd/%d", fdscript);
+                char target_path[MAXPATHLEN];
+                SSize_t len = readlink(proc_fd_path, target_path, sizeof(target_path) - 1);
```

Comment:
This should use `PerlLIO_readlink()` instead of raw `readlink()`.

The else block should be `#if` conditional on `HAS_READLINK` in addition to the `/proc/self/...` option you suggested.